src/background.ts

```diff
@@ -18,160 +17,196 @@ MainAccount m/84'/1776'/0'
+  
+      chrome.storage.local.set({ lastUsed, historyTxsId, heightsSet, txidHeight });
+      resolve({ lastUsed, historyTxsId, heightsSet, txidHeight });
+    });
+  }
+  return true;
+});
+
+
+
+async function sync(account: Account, electrum: ElectrumWS): Promise<{ 
+  lastUsed: { internal: number, external: number }, 
+  historyTxsId: Set<string>, 
+  heightsSet: Set<number>,
+  txidHeight: Map<string, number | undefined> 
+}> {
+
+  let historyTxsId: Set<string> = new Set();
+  let heightsSet: Set<number> = new Set();
+  let txidHeight: Map<string, number | undefined> = new Map();
+
+  let lastUsed = { internal: 0, external: 0 };
+  const walletChains = [0, 1];
+  for (const i of walletChains) {
+    const isInternal = i === 1;
+    let batchCount = 0;
+
+    while (true) {
+      const batch = account.deriveBatch(batchCount, GAP_LIMIT, isInternal);
+      const histories = await electrum.batchScriptGetHistory(batch);
+
+      let max = histories
+        .map((v, i) => v.length > 0 ? i : -1)
+        .reduce((a, b) => Math.max(a, b));
+      if (max >= 0) {
+        if (isInternal) {
+          lastUsed.internal = max + batchCount * GAP_LIMIT;
+        } else {
+          lastUsed.external = max + batchCount * GAP_LIMIT;
+        }
+      }
+
+
+      let flattened: GetHistoryResponse[] = histories.flat();
+      console.log(`${i}/batch(${batchCount}) ${flattened.length}`);
+
+      if (flattened.length === 0) {
+        break;
+      }
+
 
+      for (let el of flattened) {
+        let height = Math.max(el.height, 0);
+        heightsSet.add(height as number);
+        if (height === 0) {
+          txidHeight.set(el.tx_hash, undefined);
+        } else {
+          txidHeight.set(el.tx_hash, height as number);
+        }
+
+        historyTxsId.add(el.tx_hash);
+      }
+
+      batchCount += 1;
     }
-    return true;
-});
+  }
 
-async function getHistory(ws: WebSocket, address: string): Promise<Array<{ tx_hash: string }>> {
-    const scripthash = hashScriptAndReverse(address);
-    const history = await websocketRequest(ws, 'blockchain.scripthash.get_history', [scripthash]);
-    if (history.length > 0) {
-        await chrome.storage.local.set({ [HistoryKey(scripthash)]: history });
+  return { lastUsed, historyTxsId, heightsSet, txidHeight };
+}
+
+class Account {
+  node: BIP32Interface;
+  cache: Record<string, string>;
+
+  static BASE_DERIVATION_PATH = "m/84'/1776'/0'";
+  static BASE_DERIVATION_PATH_LEGACY = "m/84'/0'/0'";
+  static BASE_DERIVATIONT_PATH_TESTNET = "m/84'/1'/0'";
+
+
+  constructor(node: BIP32Interface, private network: networks.Network, private baseDerivationPath: string = Account.BASE_DERIVATION_PATH_LEGACY) {
+    this.node = node.derivePath(baseDerivationPath);
+    this.cache = {};
+  }
+
+  // Derive a range from start to end index of public keys applying the base derivation path
+  deriveBatch(start: number, end: number, isInternal: boolean): string[] {
```

Comment:
this should be based on [this](https://github.com/Blockstream/gdk/blob/9fbc7abbea98f33f698b9be9da10a452860a7b94/subprojects/gdk_rust/gdk_electrum/src/account.rs#L867-L888), but I don't fully understand the code, especially they multiply start * BATCH_SIZE etc.. therefore I changed a bit

Pretty sure there is a reason, as not all the code as been ported